src/main/java/com/lginterview/providers/CourseProviderImpl.java

```diff
@@ -0,0 +1,139 @@
+package com.lginterview.providers;
+
+import com.lginterview.common.CourseFunction;
+import com.lginterview.dto.CourseRequest;
+import com.lginterview.dto.CourseStatus;
+import com.lginterview.dto.RequestType;
+import com.lginterview.entities.Course;
+import com.lginterview.entities.User;
+import com.lginterview.exceptions.CourseDataRetrievalException;
+import com.lginterview.repositories.CourseRepository;
+import com.lginterview.repositories.UserRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.*;
+
+@Service
+public class CourseProviderImpl implements CourseProvider {
+
+    private final CourseRepository courseRepository;
+    private final UserRepository userRepository;
+    private final Map<RequestType, CourseFunction<CourseRequest, Long>> courseActions = new HashMap<>();
+
+    @Autowired
+    public CourseProviderImpl(CourseRepository courseRepository, UserRepository userRepository) {
+        this.courseRepository = courseRepository;
+        this.userRepository = userRepository;
+
+        this.courseActions.put(RequestType.INIT, (this::createCourse));
+        this.courseActions.put(RequestType.SAVE, (this::updateCourse));
+        this.courseActions.put(RequestType.FINISH, (this::closeCourse));
+    }
+
+    @Override
+    @Transactional
+    public Long modifyCourse(CourseRequest courseRequest) throws CourseDataRetrievalException {
+        if (courseRequest == null)
+            throw new IllegalArgumentException("Course Request cannot be null");
+
+        if (courseRequest.getType() == null)
+            throw new IllegalArgumentException("Course Request Type cannot be null");
+
+        if (courseActions.containsKey(courseRequest.getType()))
+            return courseActions.get(courseRequest.getType()).apply(courseRequest);
+        else
+            throw new IllegalArgumentException("Invalid Course Request type");
+    }
+
+    private Long createCourse(CourseRequest courseRequest) throws CourseDataRetrievalException {
+        Course newCourse = courseRepository.findOne(courseRequest.getCourseId());
+        if (newCourse != null)
+            return newCourse.getId();
+
+        User courseUser = userRepository.findOne(courseRequest.getUserId());
+        if (courseUser == null)
+            throw new CourseDataRetrievalException("User not found");
+
+        newCourse = new Course();
+        newCourse.setStartDate(courseRequest.getTimestamp());
+        newCourse.setCourseUser(courseUser);
+        return courseRepository.save(newCourse).getId();
```

Comment:
[MS][Minor] In case there would be a connectivity issue with DB or other error, there will be unchecked exception thrown and the clients request will be rejected with 503. 

Same applies to all other calls to DB.

Reply:
I Agree